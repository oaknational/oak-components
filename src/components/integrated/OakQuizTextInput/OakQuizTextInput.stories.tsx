import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakQuizTextInput } from "./OakQuizTextInput";

import { sizeArgTypes } from "@/storybook-helpers/sizeStyleHelpers";

const meta: Meta<typeof OakQuizTextInput> = {
  component: OakQuizTextInput,
  tags: ["autodocs"],
  title: "components/integrated/OakQuizTextInput",
  argTypes: {
    width: sizeArgTypes["$width"],
  },
  parameters: {
    controls: {
      include: ["feedback", "width", "disabled"],
    },
  },
  args: {
    placeholder: "Placeholder text",
  },
};
export default meta;

type Story = StoryObj<typeof OakQuizTextInput>;

export const Default: Story = {
  render: (args) => <OakQuizTextInput {...args} />,
};

export const WithCorrectFeedback: Story = {
  render: () => (
    <OakQuizTextInput value="A correct answer" feedback="correct" />
  ),
};

export const WithIncorrectFeedback: Story = {
  render: () => (
    <OakQuizTextInput value="An incorrect answer" feedback="incorrect" />
  ),
};

export const ResponsiveWidth: Story = {
  render: () => (
    <OakQuizTextInput
      value="An incorrect answer"
      feedback="incorrect"
      wrapperWidth={["100%", "all-spacing-22"]}
    />
  ),
};
