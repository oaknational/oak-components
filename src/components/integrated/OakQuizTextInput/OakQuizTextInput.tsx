import React from "react";

import { OakTextInput, OakTextInputProps } from "@/components/ui";

type OakQuizTextInputProps = Omit<
  OakTextInputProps,
  "validity" | "iconName" | "iconAlt" | "isTrailingIcon"
> & {
  /**
   * Alters the appearance of the input to indicate whether or not a correct answer was given.
   * Also sets the input to read-only.
   */
  feedback?: "correct" | "incorrect" | null;
};

export const OakQuizTextInput = ({
  feedback,
  readOnly,
  ...props
}: OakQuizTextInputProps) => {
  let validity: OakTextInputProps["validity"];
  let iconName: OakTextInputProps["iconName"];
  let iconAlt: OakTextInputProps["iconAlt"] = undefined;

  switch (feedback) {
    case "correct":
      validity = "valid";
      iconName = "tick";
      iconAlt = "Correct";
      break;
    case "incorrect":
      validity = "invalid";
      iconName = "cross";
      iconAlt = "Incorrect";
      break;
  }

  return (
    <OakTextInput
      {...props}
      validity={validity}
      iconName={iconName}
      iconAlt={iconAlt}
      readOnly={readOnly || !!feedback}
      isTrailingIcon
    />
  );
};
