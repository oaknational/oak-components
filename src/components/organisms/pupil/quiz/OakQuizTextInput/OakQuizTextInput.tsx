import React from "react";

import { OakTextInput, OakTextInputProps } from "@/components/molecules";

type OakQuizTextInputProps = Omit<
  OakTextInputProps,
  "validity" | "iconName" | "iconAlt" | "isTrailingIcon"
> & {
  /**
   * Alters the appearance of the input to indicate whether or not a correct answer was given.
   * Also sets the input to read-only.
   */
  feedback?: "correct" | "incorrect" | null;
};

/**
 * A text input for a free-text question in a quiz
 */
export const OakQuizTextInput = ({
  feedback,
  readOnly,
  ...props
}: OakQuizTextInputProps) => {
  let validity: OakTextInputProps["validity"];
  let iconName: OakTextInputProps["iconName"];
  let iconAlt: OakTextInputProps["iconAlt"] = undefined;

  switch (feedback) {
    case "correct":
      validity = "valid";
      iconName = "tick";
      iconAlt = "Correct";
      break;
    case "incorrect":
      validity = "invalid";
      iconName = "cross";
      iconAlt = "Incorrect";
      break;
  }

  return (
    <OakTextInput
      {...props}
      validity={validity}
      iconName={iconName}
      iconAlt={iconAlt}
      readOnly={readOnly || !!feedback}
      isTrailingIcon
    />
  );
};
