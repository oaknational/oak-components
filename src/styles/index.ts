export * from "@/styles/theme/color";
