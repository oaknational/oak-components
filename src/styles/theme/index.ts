export { oakColorTokens, oakUiRoleTokens } from "./color";

export type {
  OakColorToken,
  OakCombinedColorToken,
  OakUiRoleToken,
  OakColorFilterToken,
} from "./color";

export {
  oakAllSpacingTokens,
  oakInnerPaddingTokens,
  oakSpaceBetweenTokens,
} from "./spacing";

export type {
  OakAllSpacingToken,
  OakCombinedSpacingToken,
  OakInnerPaddingToken,
  OakSpaceBetweenToken,
} from "./spacing";

export { oakDropShadowTokens } from "./dropShadow";
export type { OakDropShadowToken } from "./dropShadow";

export { oakOpacityTokens } from "./opacity";
export type { OakOpacityToken } from "./opacity";

export { oakTransitionTokens } from "./transitions";
export type { OakTransitionToken } from "./transitions";

export { oakBorderRadiusTokens, oakBorderWidthTokens } from "./borders";
export type { OakBorderRadiusToken, OakBorderWidthToken } from "./borders";

export type { OakTheme } from "./theme";

export { oakFontTokens, oakFontSizeTokens } from "./typography";
export type { OakFontToken, OakFontSizeToken } from "./typography";

export { oakZIndexTokens } from "./zIndex";

export { oakDefaultTheme } from "./default.theme";
