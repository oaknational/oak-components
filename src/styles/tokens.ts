export const oakAllSpacingPx = {
  "all-spacing-1": 4,
  "all-spacing-2": 8,
  "all-spacing-3": 12,
  "all-spacing-4": 16,
  "all-spacing-5": 20,
  "all-spacing-6": 24,
  "all-spacing-7": 32,
  "all-spacing-8": 40,
  "all-spacing-9": 48,
  "all-spacing-10": 56,
  "all-spacing-11": 64,
  "all-spacing-12": 72,
  "all-spacing-13": 80,
  "all-spacing-14": 92,
  "all-spacing-15": 100,
  "all-spacing-16": 120,
};

export type OakAllSpacing = keyof typeof oakAllSpacingPx;

export const oakInnerPaddingAllSpacing = {
  "inner-padding-xs": "all-spacing-2",
  "inner-padding-s": "all-spacing-3",
  "inner-padding-m": "all-spacing-4",
  "inner-padding-l": "all-spacing-5",
  "inner-padding-xl": "all-spacing-6",
};

export type OakInnerPadding = keyof typeof oakInnerPaddingAllSpacing;

export const oakSpaceBetweenAllSpacing = {
  "space-between-ssx": "all-spacing-2",
  "space-between-xs": "all-spacing-3",
  "space-between-s": "all-spacing-4",
  "space-between-m": "all-spacing-6",
  "space-between-m2": "all-spacing-7",
  "space-between-l": "all-spacing-9",
  "space-between-xl": "all-spacing-10",
  "space-between-xxl": "all-spacing-12",
  "space-between-xxxl": "all-spacing-13",
};

export type OakSpaceBetween = keyof typeof oakSpaceBetweenAllSpacing;

export const oakAllColorsHex = {
  white: "#ffffff",
  grey20: "#f2f2f2",
  grey30: "#e4e4e4",
  grey40: "#cacaca",
  grey50: "#808080",
  grey60: "#575757",
  grey70: "#2d2d2d",
  black: "#222222",
  oakGreen: "#287c34",
  mint: "bef2bd",
  mint30: "ebfbeb",
  mint50: "dff9de",
  aqua: "#b0e2de",
  aqua30: "#e7f6f5",
  aqua50: "#cee7e5",
  lavender: "#a0b6f2",
  lavender30: "#e3e9fb",
  lavender50: "#c6d1ef",
  pink: "#deb7d5",
  pink30: "#f5e9f2",
  pink50: "#e5d1e0",
  lemon: "#ffe555",
  lemon30: "#fff7cc",
  lemon50: "#f6e8a0",
  amber: "#ff934e",
  amber30: "#ffdfca",
  amber50: "#ffc8a6",
  red: "#dd0035",
  navy: "#0d24c4",
  blue: "#374cf1",
  magenta: "#d02aa7",
  purple: "#845ad9",
  teal: "#037b7d",
};

export type OakColor = keyof typeof oakAllColorsHex;

export const oakColorTokens = [
  "text-primary",
  "text-subdued",
  "text-error",
  "text-disabled",
  "text-link-active",
  "text-link-hover",
  "text-link-visited",
  "text-inverted",
  "text-success",
  "text-warning",
  "bg-primary",
  "bg-neutral",
  "bg-neutral-50",
  "bg-neutral-30",
  "bg-btn-primary",
  "bg-btn-primary-hover",
  "bg-btn-primary-disabled",
  "bg-btn-secondary",
  "bg-btn-secondary-hover",
  "bg-btn-secondary-disabled",
  "bg-icon",
  "bg-icon-hover",
  "bg-decorative1â€“main",
  "bg-decorative1-subdued",
  "bg-decorative1-very subdued",
  "bg-decorative2-main",
  "bg-decorative2-subdued",
  "bg-decorative2-very-subdued",
  "bg-decorative3",
  "bg-decorative3-50",
  "bg-decorative3-30",
  "bg-decorative4",
  "bg-decorative4-50",
  "bg-decorative4-30",
  "bg-decorative5",
  "bg-decorative5-50",
  "bg-decorative5-30",
  "bg-decorative6",
  "bg-decorative6-50",
  "bg-decorative6-30",
  "border-primary",
  "border-inverted",
  "border-neutral",
  "border-decorative1",
  "border-decorative2",
  "border-decorative3",
  "border-decorative4",
  "border-decorative5",
  "border-decorative6",
] as const;

export type OakColorToken = (typeof oakColorTokens)[number];
