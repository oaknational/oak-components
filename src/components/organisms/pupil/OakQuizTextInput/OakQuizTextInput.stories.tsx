import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakQuizTextInput } from "./OakQuizTextInput";

import { sizeArgTypes } from "@/storybook-helpers/sizeStyleHelpers";

const meta: Meta<typeof OakQuizTextInput> = {
  component: OakQuizTextInput,
  tags: ["autodocs"],
  title: "components/organisms/pupil/OakQuizTextInput",
  argTypes: {
    wrapperWidth: sizeArgTypes["$width"],
  },
  parameters: {
    controls: {
      include: ["feedback", "wrapperWidth", "disabled"],
    },
  },
  args: {
    placeholder: "Placeholder text",
  },
};
export default meta;

type Story = StoryObj<typeof OakQuizTextInput>;

export const Default: Story = {
  render: (args) => <OakQuizTextInput {...args} />,
};

export const WithCorrectFeedback: Story = {
  render: () => (
    <OakQuizTextInput value="A correct answer" feedback="correct" />
  ),
};

export const WithIncorrectFeedback: Story = {
  render: () => (
    <OakQuizTextInput value="An incorrect answer" feedback="incorrect" />
  ),
};

export const ResponsiveWidth: Story = {
  render: (args) => (
    <OakQuizTextInput
      wrapperWidth={["100%", "all-spacing-20"]}
      feedback={args.feedback}
      defaultValue={args.defaultValue}
    />
  ),
  args: {
    feedback: "incorrect",
    defaultValue:
      "The pupil's answer which is longer than the width of the input",
  },
  parameters: {
    controls: {
      include: ["feedback"],
    },
  },
};
