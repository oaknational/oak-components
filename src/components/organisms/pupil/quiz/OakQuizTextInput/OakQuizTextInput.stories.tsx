import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakQuizTextInput } from "./OakQuizTextInput";

import { sizeArgTypes } from "@/storybook-helpers/sizeStyleHelpers";

const meta: Meta<typeof OakQuizTextInput> = {
  component: OakQuizTextInput,
  tags: ["autodocs"],

  argTypes: {
    wrapperWidth: sizeArgTypes["$width"],
  },
  parameters: {
    controls: {
      include: ["feedback", "wrapperWidth", "disabled"],
    },
  },
  args: {
    placeholder: "Placeholder text",
  },
  render: (args) => <OakQuizTextInput {...args} />,
};
export default meta;

type Story = StoryObj<typeof OakQuizTextInput>;

export const Default: Story = {};

export const WithCorrectFeedback: Story = {
  args: {
    value: "A correct answer",
    feedback: "correct",
  },
};

export const WithIncorrectFeedback: Story = {
  args: {
    value: "An incorrect answer",
    feedback: "incorrect",
  },
};

export const Highlighted: Story = {
  args: {
    isHighlighted: true,
  },
};

export const ResponsiveWidth: Story = {
  args: {
    wrapperWidth: ["100%", "all-spacing-20"],
    feedback: "incorrect",
    defaultValue:
      "The pupil's answer which is longer than the width of the input",
  },
  parameters: {
    controls: {
      include: ["feedback"],
    },
  },
};
