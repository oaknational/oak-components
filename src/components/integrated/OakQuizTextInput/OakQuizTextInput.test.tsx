import React from "react";
import "@testing-library/jest-dom";
import { create } from "react-test-renderer";
import { ThemeProvider } from "styled-components";

import { OakQuizTextInput } from "./OakQuizTextInput";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { oakDefaultTheme } from "@/styles";

describe("OakQuizTextInput", () => {
  it("renders", () => {
    const { getByTestId } = renderWithTheme(
      <OakQuizTextInput
        defaultValue="An answer to a question"
        data-testid="quiz-input"
      />,
    );

    expect(getByTestId("quiz-input")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = create(
      <ThemeProvider theme={oakDefaultTheme}>
        <OakQuizTextInput defaultValue="An answer to a question" />,
      </ThemeProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('sets the input to read-only when feedback is "correct"', () => {
    const { getByRole, getByAltText } = renderWithTheme(
      <OakQuizTextInput
        defaultValue="An answer to a question"
        feedback="correct"
      />,
    );

    expect(getByRole("textbox")).toHaveAttribute("readonly");
    expect(getByAltText("Correct")).toBeInTheDocument();
  });

  it('sets the input to read-only when feedback is "incorrect"', () => {
    const { getByRole, getByAltText } = renderWithTheme(
      <OakQuizTextInput
        defaultValue="An answer to a question"
        feedback="incorrect"
      />,
    );

    expect(getByRole("textbox")).toHaveAttribute("readonly");
    expect(getByAltText("Incorrect")).toBeInTheDocument();
  });
});
