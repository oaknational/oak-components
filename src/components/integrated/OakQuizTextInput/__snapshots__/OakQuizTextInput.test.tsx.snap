// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakQuizTextInput matches snapshot 1`] = `
[
<<<<<<< HEAD
  <div
    className="sc-aXZVg sc-gEvEer sc-koXPp dyiMma ftzkJO iiiHCi"
    onClick={[Function]}
  >
    <input
      className="sc-eeDRCY dgoclW"
=======
  .c0 {
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  padding-left: 1.25rem;
  padding-right: 1.25rem;
  color: #222222;
  background: #ffffff;
  border: 0.125rem solid;
  border-color: #222222;
  border-radius: 0.25rem;
  font-family: Lexend,sans-serif;
}

.c0:hover {
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
}

.c3 {
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
  border: 0;
  border-radius: 0;
  border-color: transparent;
  box-shadow: none;
  font-family: inherit;
  background: transparent;
  outline: none;
  color: inherit;
  padding-top: 1.25rem;
  padding-bottom: 1.25rem;
  height: 4.5rem;
}

.c3::-webkit-input-placeholder {
  color: #575757;
}

.c3::-moz-placeholder {
  color: #575757;
}

.c3:-ms-input-placeholder {
  color: #575757;
}

.c3::placeholder {
  color: #575757;
}

.c3::-webkit-search-decoration,
.c3::-webkit-search-cancel-button,
.c3::-webkit-search-results-button,
.c3::-webkit-search-results-decoration {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c3:disabled {
  cursor: not-allowed;
}

.c2 {
  background: #ffffff;
}

.c2:hover {
  cursor: text;
}

.c2:focus-within {
  box-shadow: 0 0 0 0.125rem rgba(255,229,85,100%),0 0 0 0.3rem rgba(87,87,87,100%);
}

@media (max-width:750px) {
  .c3 {
    font-size: 16px;
  }
}

@media (hover:hover) {
  .c2:hover:not(:focus-within) {
    background: #f2f2f2;
  }
}

<div
    className="c0 c1 c2"
    onClick={[Function]}
  >
    <input
      className="c3"
>>>>>>> main
      defaultValue="An answer to a question"
      onFocus={[Function]}
      readOnly={false}
      type="text"
    />
  </div>,
  ",",
]
`;
